### PR TITLE
`ignore_above` default to 8191 for `logsdb`

### PR DESCRIPTION
**Pull request by salvatore-campagna**
In LogsDB we would like to use a default value of `8191` for the index-level setting
`index.mapping.ignore_above`. The value for `ignore_above` is the _character count_,
but Lucene counts bytes. Here we set the limit to `32766 / 4 = 8191` since UTF-8
characters may occupy at most 4 bytes.

### CODE REVIEW

**Comment by dnhatn** [comment]
The changes look good, but I think 1024 is too small to be the default.

**Comment by salvatore-campagna** [comment]
@felixbarny @andrewkroh We would like to know if this value is good for our integrations.
Especially for fields that **do not** have a value set at mapping level, an index-level setting would result in applying a possibly unwanted limit. For fields which might include stack traces we might end-up missing the ability to search on them.

**Comment by felixbarny** [comment]
1024 is the default that is used in ECS. I'm not sure where this exact number was determined and why we haven't chosen a slightly larger default. The intention behind it is to avoid document rejections when running into the Lucene's byte-length limit of 32kb. I think we can choose a generic default for LogsDB that would prevent us running into that limit (with any unicode character) but that's bigger than the ECS limit of 1024.

While I think it makes sense to have a default value for `ignore_above` in LogsDB to avoid rejections, I think we should discuss whether there are better alternatives to `ignore_above`. For example, we could ignore values only if they would actually exceed the hard-limit in Lucene. Or instead of ignoring them entirely, we could truncate the values. There are some discussions in this issue that we may want to revise: https://github.com/elastic/elasticsearch/issues/60329

What seems missing in your PR is to actually use the index setting in keyword field mappers and updating the docs to describe that there's a way to set an index-level default, similar to what we do with [`ignore_malformed`](https://www.elastic.co/guide/en/elasticsearch/reference/current/ignore-malformed.html#ignore-malformed-setting).

**Comment by salvatore-campagna** [comment]
> 1024 is the default that is used in ECS. I'm not sure where this exact number was determined and why we haven't chosen a slightly larger default. The intention behind it is to avoid document rejections when running into the Lucene's byte-length limit of 32kb. I think we can choose a generic default for LogsDB that would prevent us running into that limit (with any unicode character) but that's bigger than the ECS limit of 1024.
> 
> While I think it makes sense to have a default value for `ignore_above` in LogsDB to avoid rejections, I think we should discuss whether there are better alternatives to `ignore_above`. For example, we could ignore values only if they would actually exceed the hard-limit in Lucene. Or instead of ignoring them entirely, we could truncate the values. There are some discussions in this issue that we may want to revise: #60329
> 
> What seems missing in your PR is to actually use the index setting in keyword field mappers and updating the docs to describe that there's a way to set an index-level default, similar to what we do with [`ignore_malformed`](https://www.elastic.co/guide/en/elasticsearch/reference/current/ignore-malformed.html#ignore-malformed-setting).

The `ignore_above` index setting was already introduced by https://github.com/elastic/elasticsearch/pull/113121.
In this PR we just change the default value from `Inter.MAX_VALUE` to `1024` (or any other suitable value we decide) for LogsDB.

**Comment by salvatore-campagna** [comment]
@felixbarny @andrewkroh we need feedback here, possibly before GA release.

**Comment by felixbarny** [comment]
+1 on having a default value for `ignore_above` for now that's potentially higher than 1024 but still guarantees that we're not hitting the hard limit of 32kb in Lucene.

What's the highest value of ignore_above that would guarantee us to be under the hard-limit in Lucene? How are we encoding chars and what's the highest number of bytes per char? Is there any other static overhead in the encoding? How are we handling unicode code points that consist of multiple characters?

**Comment by salvatore-campagna** [comment]
> +1 on having a default value for `ignore_above` for now that's potentially higher than 1024 but still guarantees that we're not hitting the hard limit of 32kb in Lucene.
> 
> What's the highest value of ignore_above that would guarantee us to be under the hard-limit in Lucene? How are we encoding chars and what's the highest number of bytes per char? Is there any other static overhead in the encoding? How are we handling unicode code points that consist of multiple characters?


From our documentation:
```
NOTE: The value for `ignore_above` is the _character count_, but Lucene counts
bytes. If you use UTF-8 text with many non-ASCII characters, you may want to
set the limit to `32766 / 4 = 8191` since UTF-8 characters may occupy at most
4 bytes.
```

In Lucene the hard limit is 32766 **bytes**.

The highest value for `ignore_above` that guarantees being under the Lucene limit, accounting for the worst-case 4-byte encoding, is 8191.

Applying the value will depend on encoding and character count will match bytes count only if all characters are encoded using one byte. In UTF-8 encoding might result in 1 to 4 bytes per character if we consider also non-ascii characters but I guess for logging purposes we can safely assume that we are dealing with ASCII characters.

(AFAIK Elasticsearch uses UTF-8 encoding for strings).

About the static overhead I don't see any issue...keywords and text are normally converted to arrays of bytes.

**Comment by felixbarny** [comment]
> The highest value for ignore_above that guarantees being under the Lucene limit, accounting for the worst-case 4-byte encoding, is 8191.

Let's set the default limit to 8191 then.

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/index/IndexSettings.java [hunk 1]
```diff
@@ -713,14 +713,20 @@ public Iterator<Setting<?>> settings() {
...
-        Integer.MAX_VALUE,
-        0,
-        Property.IndexScope,
-        Property.ServerlessPublic
-    );
+
+    public static final Setting<Integer> IGNORE_ABOVE_SETTING = Setting.intSetting("index.mapping.ignore_above", settings -> {
+        if (IndexSettings.MODE.get(settings) == IndexMode.LOGSDB
+            && IndexMetadata.SETTING_INDEX_VERSION_CREATED.get(settings).onOrAfter(IndexVersions.ENABLE_IGNORE_ABOVE_LOGSDB)) {
+            return "8191";
```
Comment (kkrik-es) [inline server/src/main/java/org/elasticsearch/index/IndexSettings.java:725]: How can we override this for an index with logsdb mode?

Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/index/IndexSettings.java:725]: I see, this is the default value.. Let's add a comment above, or move it to a static helper function for clarity.

Comment (salvatore-campagna) [reply server/src/main/java/org/elasticsearch/index/IndexSettings.java:725]: The

```
settings -> { ... }
```

lambda just determines the default value if no explicit value is provided for the setting.
I will extract the lambda in a method with a descriptive name.

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

**Comment by salvatore-campagna** [comment]
@martijnvg @felixbarny I need an approval if we ant to merge this

**Review by felixbarny — APPROVED** [review]

**Review by martijnvg — APPROVED** [review]
LGTM

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

**Comment by lkts** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[](https://github.com/elastic/elasticsearch/pull/115373)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

**Comment by lkts** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.16|[](https://github.com/elastic/elasticsearch/pull/115451)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

**Pull request merged**